http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/ResponseHandler.java

```diff
@@ -106,14 +109,52 @@ protected void channelRead0(ChannelHandlerContext channelContext, HttpObject msg
 
                 fullContent = ((FullHttpResponse) msg).content();
                 ByteBuffer bb = copyToByteBuffer(fullContent);
-                requestContext.handler().onStream(new FullResponseContentPublisher(channelContext, bb, ef));
-                finalizeResponse(requestContext, channelContext);
+                requestContext.handler().onStream(new DataCountingPublisher(channelContext,
+                                                                            new FullResponseContentPublisher(channelContext,
+                                                                                                             bb, ef)));
+
+                Optional<IOException> contentLengthFailure = validateResponseContentLength(channelContext);
+                if (contentLengthFailure.isPresent()) {
+                    exceptionCaught(channelContext, contentLengthFailure.get());
+                } else {
+                    finalizeResponse(requestContext, channelContext);
+                }
             } finally {
                 Optional.ofNullable(fullContent).ifPresent(ByteBuf::release);
             }
         }
     }
 
+    private Long responseContentLength(HttpResponse response) {
+        String length = response.headers().get("content-length");
+        if (length == null) {
+            return null;
+        }
+
+        return Long.parseLong(length);
+    }
+
+    private static Optional<IOException> validateResponseContentLength(ChannelHandlerContext ctx) {
+        Long contentLengthHeader = ctx.channel().attr(RESPONSE_CONTENT_LENGTH).get();
+        Long actualContentLength = ctx.channel().attr(RESPONSE_DATA_READ).get();
+
+        if (contentLengthHeader == null) {
+            return Optional.empty();
+        }
+
+        if (actualContentLength == null) {
+            actualContentLength = 0L;
+        }
+
+        if (actualContentLength.equals(contentLengthHeader)) {
+            return Optional.empty();
+        }
+
+        return Optional.of(new IOException("Response had content-length of " + contentLengthHeader + " bytes, but only received "
+                                           + actualContentLength + " before the connection was closed."));
```

Comment:
nit: "but only received X **bytes** before"

nit: should we emit a different/special message if we received 0 bytes?

nit: should we wrap the `IOException` inside of Netty's `PrematureChannelClosureException`?

Reply:
It would need to be the other way around: IOException on the outside, so it's seen a retryable. PrematureChannelClosureException is a RuntimeException. Does that change your mind on it? Should I use that inside?

Reply:
Doesn't our default retry logic check 1-level of causes?

What happens if Netty does actually throw a `PrematureChannelClosureException` itself? E.g., [here](https://github.com/netty/netty/blob/f1eb920467ebad658fc0a108f73a98017fa28612/codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectDecoder.java#L482). Do we retry on that? It seems like we would want to.

Reply:
It does check 1 level of causes, but I was under the impression that the contract for `ExecutableHttpRequest.call` is that only IOException will be retried, not necessarily one level of exception. I don't know if I have the wrong impression, or if there's further wrapping that happens upstream before the retry logic that necessitates checking 1 level deeper.